### PR TITLE
feat(): add validating webhook (delete) to prevent vpn key rotation config delete

### PR DESCRIPTION
**Pull request by rahulsawra98**
This PR adds w delete validating webhook!

# Description

Fixes #

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit test cases.

## Does this PR introduce a breaking change for other components like worker-operator?

```release-note

```

### CODE REVIEW

**Review by PuranAdhikari — APPROVED** [review]

**Comment by rahulsawra98** [comment]
@PuranAdhikari re-requesting review!

File: config/events/controller.yaml [hunk 1]
```diff
@@ -515,3 +515,9 @@ events:
     type: Normal
     reportingController: controller
     message: Certificates to be renewed Now!
+  - name: VPNKeyRotationConfigConfigDeletedForcefully
+    reason: VPNKeyRotationConfigConfigDeletedForcefully
```
Comment (rajendra-avesha) [inline config/events/controller.yaml:519]: Duplicate word ConfigConfig

File: config/events/events_config_map.yaml [hunk 2]
```diff
@@ -92,4 +92,5 @@ data:
       - VPNKeyRotationStart
       - VPNKeyRotationConfigUpdated
       - CertificateJobCreationFailed
-      - CertificatesRenewNow
\ No newline at end of file
+      - CertificatesRenewNow
+      - VPNKeyRotationConfigConfigDeletedForcefully
```
Comment (rajendra-avesha) [inline config/events/events_config_map.yaml:96]: Duplicate word ConfigConfig

File: events/events_generated.go [hunk 3]
```diff
@@ -702,92 +702,101 @@ var EventsMap = map[events.EventName]*events.EventSchema{
 		ReportingController: "controller",
 		Message:             "Certificates to be renewed Now!",
 	},
+	"VPNKeyRotationConfigConfigDeletedForcefully": {
+		Name:                "VPNKeyRotationConfigConfigDeletedForcefully",
```
Comment (rajendra-avesha) [inline events/events_generated.go:706]: Duplicate word ConfigConfig

File: config/events/controller.yaml [hunk 4]
```diff
@@ -515,3 +515,9 @@ events:
...
+    action: DeleteVPNKeyRotationConfig
+    type: Warning
+    reportingController: controller
+    message: VPNKeyRotationConfig config got deleted forcefully.
```
Comment (rajendra-avesha) [inline config/events/controller.yaml:523]: Please note the CR is not deleted here please change the message appropriately

File: config/events/controller.yaml [hunk 5]
```diff
@@ -515,3 +515,9 @@ events:
     type: Normal
     reportingController: controller
     message: Certificates to be renewed Now!
+  - name: VPNKeyRotationConfigConfigDeletedForcefully
```
Comment (rajendra-avesha) [inline config/events/controller.yaml:518]: Can we change it IllegalVPNKeyRotationConfigDelete

File: config/events/events_config_map.yaml (hunk 2 above)
Comment (rajendra-avesha) [inline config/events/events_config_map.yaml:96]: Can we add an event for successful delete of VPNKeyRotationConfig by controller

File: service/vpn_key_rotation_webhook_validation_test.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
+package service
+
+import (
+	"context"
```
Comment (rajendra-avesha) [inline service/vpn_key_rotation_webhook_validation_test.go:4]: Please add ITs for the create and delete webhooks which should check for the events too

**Review by rajendra-avesha — CHANGES_REQUESTED** [review]

File: config/events/events_config_map.yaml (hunk 2 above)
Comment (rahulsawra98) [reply config/events/events_config_map.yaml:96]: Its not done by controller. Its the owner ref.

File: service/vpn_key_rotation_webhook_validation.go [hunk 7]
```diff
@@ -16,3 +19,26 @@ func ValidateVpnKeyRotationCreate(ctx context.Context, r *controllerv1alpha1.Vpn
...
+	}
+	if found && slice.ObjectMeta.DeletionTimestamp.IsZero() {
+		//Load Event Recorder with project name, vpnkeyrotation(slice) name and namespace
+		eventRecorder := util.CtxEventRecorder(ctx).
+			WithProject(util.GetProjectName(r.Namespace)).
+			WithNamespace(r.Namespace).
+			WithSlice(r.Name)
+		//Register an event for worker slice config deleted forcefully
+		util.RecordEvent(ctx, eventRecorder, r, slice, events.EventIllegalVPNKeyRotationConfigDelete)
+		return fmt.Errorf("sliceconfig %s not allowed to delete unless slice is deleted", r.Name)
```
Comment (PuranAdhikari) [inline service/vpn_key_rotation_webhook_validation.go:40]: Can we change this message a bit. 
`VPN Key is not allowed to delete unless sliceconfig is deleted.`

File: service/vpn_key_rotation_webhook_validation_test.go (hunk 6 above)
Comment (rajendra-avesha) [reply service/vpn_key_rotation_webhook_validation_test.go:4]: Where are ITs?

File: controllers/controller/vpnkey_rotation_controller_test.go [hunk 8]
```diff
@@ -283,16 +397,203 @@ var _ = Describe("VpnKeyRotation Controller", Ordered, func() {
...
+			// it should delete sliceconfig
+			Expect(k8sClient.Delete(ctx, slice)).Should(Succeed())
+			Expect(k8sClient.Delete(ctx, cluster1)).Should(Succeed())
+			Expect(k8sClient.Delete(ctx, cluster2)).Should(Succeed())
+			Expect(k8sClient.Delete(ctx, cluster3)).Should(Succeed())
+
+			Expect(k8sClient.Delete(ctx, project)).Should(Succeed())
+
+		})
+		It("Should not allow deleting vpnkeyrotation config, if slice is present and raise an event", func() {
```
Comment (rajendra-avesha) [inline controllers/controller/vpnkey_rotation_controller_test.go:567]: Can we add one case +ve case where delete slice config should delete the VPN config too

**Pull request merged**